contracts/RewardController.sol

```diff
@@ -0,0 +1,252 @@
+                totalAllocPoint: totalAllocPoint
+            }));
+        }
+
+        vaultInfo[_vault].allocPoint = _allocPoint;
+    }
+
+    /**
+    * @notice Safe HAT transfer function, transfer rewards from the contract only if there are enough
+    * rewards available.
+    * @param _to The address to transfer the reward to
+    * @param _amount The amount of rewards to transfer
+    * @param _vault The vault address
+   */
+    function safeTransferReward(address _to, uint256 _amount, address _vault) internal {
+        if (rewardToken.balanceOf(address(this)) < _amount)
+            revert NotEnoughRewardsToTransferToUser();
+            
+        rewardToken.safeTransfer(_to, _amount);
+
+        emit SafeTransferReward(_to, _vault, _amount, address(rewardToken));
+    }
+
+    /**
+    * @notice Update the vault's rewardPerShare, not more then once per block
+    * @param _vault The vault's address
+    */
+    function updateVault(address _vault) public {
+        VaultInfo storage vault = vaultInfo[_vault];
+        uint256 lastRewardBlock = vault.lastRewardBlock;
+        if (lastRewardBlock == 0) {
+            vaultInfo[_vault].lastRewardBlock = block.number > startBlock ? block.number : startBlock;
+            return;
+        }
+        if (block.number <= lastRewardBlock) {
+            return;
+        }
+
+        vault.lastRewardBlock = block.number;
+
+        uint256 totalShares = IERC20Upgradeable(_vault).totalSupply();
+
+        if (totalShares != 0) {
+            uint256 reward = getVaultReward(_vault, lastRewardBlock);
+            vault.rewardPerShare += (reward * 1e12 / totalShares);
+        }
+
+        if (globalVaultsUpdates.length != 0) {
+            vaultInfo[_vault].lastProcessedTotalAllocPoint = globalVaultsUpdates.length - 1;
+        }
+    }
+
+    /**
+     * @notice Called by owner to set reward multipliers
+     * @param _rewardPerEpoch reward multipliers
+    */
+    function setRewardPerEpoch(uint256[24] memory _rewardPerEpoch) external onlyOwner {
+        rewardPerEpoch = _rewardPerEpoch;
+        emit SetRewardPerEpoch(_rewardPerEpoch);
+    }
+
+    function _updateVaultBalance(
+        address _vault,
+        address _user,
+        uint256 _sharesChange,
+        bool _isDeposit
+    ) internal {
+        updateVault(_vault);
+
+        uint256 userShares = IERC20Upgradeable(_vault).balanceOf(_user);
+        uint256 rewardPerShare = vaultInfo[_vault].rewardPerShare;
+        unclaimedReward[_vault][_user] += userShares * rewardPerShare / 1e12 - rewardDebt[_vault][_user];
+
+        if (_sharesChange != 0) {
+            if (_isDeposit) {
+                userShares += _sharesChange;
+            } else {
+                userShares -= _sharesChange;
+            }
+        }
+        rewardDebt[_vault][_user] = userShares * rewardPerShare / 1e12;
+    }
+
+    function updateVaultBalance(address _user, uint256 _sharesChange, bool _isDeposit) external {
```

Comment:
could we renmae this perhaps to "updateUserRewardBalance" or something?

Reply:
idk

Reply:
Maybe _commitUserBalance or something like that?

Reply:
sounds good

Reply:
https://github.com/hats-finance/hats-contracts/pull/321